docs/design/process_group_status.md

```diff
@@ -0,0 +1,108 @@
+operator will create one. The term "process" will refer to a single fdbserver
+process managed by the operator. In configurations where we are running multiple
+storage servers on a single disk, the pod will be represented by a single
+process group, with multiple fdbserver processes within it.
+
+We will have a map of `processGroups` in the status. The keys in the map will
+be the instance ID, which will be renamed to process group ID in the future. The
+values in the map will be `processGroupStatus` entries, which will contain the
+following fields:
+
+*	`processClass`: The process class the process group has.
+*	`addresses`: A list of addresses the process group has been known to have.
+*	`remove`: Whether the process group has been marked for removal.
+*	`excluded`: Whether the process group has been fully excluded.
+*	`conditions`: A list of degraded conditions that the process group is in.
+	This can include `notConnecting`, `incorrectPodSpec`, `incorrectConfigMap`,
+	`incorrectCommandLine`, `podFailing`, `missingPod`, `missingPvc`, and 
+	`missingService`. Each entry will include the condition, and the timestamp
+	when we first observed the condition.
+
+The `UpdateStatus` action will fetch the existing resources and add any missing
+entries into the `processGroups` map. Any entries in the current map that do not
+have existing resources will be left as-is. If the current address is different
+from the last known address, and the process group is marked for removal, we
+will add the current address to the `addresses` list. If the current address is
+different from the last known address, and the process group is not marked for
+removal, we will replace the address list with the current address. The
+`UpdateStatus` method will also set the conditions on the process group based
+on the database status and the results of the calls to the resource list APIs.
+If a pod does not currently have an IP address, we will leave the address list
+unmodified, allowing us to continue to take actions that require an IP address,
+such as excluding processes, by optimistically re-using the previous address.
+
+The logic for adding new process groups will be extracted into its own action,
+which will compare the current process group map with the desired counts and
+add new entries to the map. After that action, we will run separate actions for
+creating new services, new PVCs, and new pods, in that order. Each of those
+actions will target a single resource and will update the status to reflect the
+new conditions.
+
+The exclusion action will use the `processGroups` as its source of truth, and
+will exclude all of the addresses for process groups marked for removal. If a
+process group has no addresses, then we will not be able to complete the
+exclusion action. The inclusion action will include all of the addresses for the
+process groups, and will then remove those entries from the map.
```

Comment:
Can you elaborate a bit on this process? I think I understand it, but just want to be totally clear. When a process group is marked for inclusion, you will run an `include` command on FDB with the know addresses for that process group, but you'll then remove the process group entirely from the map? 

Are there not scenarios where we want to not remove a process group entirely from the cluster, and instead just exclude then re-include it?

Reply:
We have scenarios like that, but it's not currently supported by the operator. The operator only excludes / includes processes as part of removing them.